### ISSUE TITLE
New check: JavadocSeeTagOrder

### ISSUE THREAD

**Issue by gianmarcoschifone** [checkstyle/checkstyle#19949]
# Description
After https://github.com/checkstyle/checkstyle/issues/19914

From the Javadoc style guide:

> Multiple `@see` tags should be ordered as follows, which is roughly the same order as their arguments are searched for by javadoc, basically from nearest to farthest access, from least-qualified to fully-qualified.
>
> Notice the methods and constructors are in "telescoping" order, which means the "no arg" form first, then the "1 arg" form, then the "2 arg" form, and so forth.

Multiple `@see` tags should follow a predictable order.

The check should validate that multiple `@see` tags are ordered by structure:

1. local members first
2. class references after local members
3. qualified class references after simple class references
4. package references last
5. inside each class/member group:
   - fields first
   - constructors second
   - methods third
6. overloaded constructors and methods should be ordered by number of parameters:
   - no-arg first
   - one-arg second
   - two-arg third
   - and so on
7. overloaded methods with the same name should be grouped together

# Config

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocSeeTagOrder"/>
    </module>
</module>
```

# Examples

```java
public class Test {

    /**
     * Valid: local field, local constructors, local methods,
     * simple class, simple class members, qualified class,
     * qualified class members, package.
     *
     * @see #field
     * @see #Test()
     * @see #Test(String)
     * @see #getName()
     * @see #getName(String)
     * @see OtherClass
     * @see OtherClass#field
     * @see OtherClass#OtherClass()
     * @see OtherClass#OtherClass(String)
     * @see OtherClass#getName()
     * @see OtherClass#getName(String)
     * @see java.util.List
     * @see java.util.List#size()
     * @see java.util
     */
    private String field;

    /**
     * Invalid: local method appears before local constructor.
     *
     * @see #field
     * @see #getName()
     * @see #Test() // violation
     */
    public Test() {
    }

    /**
     * Invalid: overloaded constructors are not in telescoping order.
     * No-arg constructor should come before one-arg constructor.
     *
     * @see #Test(String)
     * @see #Test() // violation
     * @see #Test(String, int)
     */
    public Test(String value) {
    }

    /**
     * Invalid: overloaded methods with the same name are not grouped together.
     *
     * @see #getName()
     * @see #setName(String)
     * @see #getName(String) // violation
     */
    public void getName() {
    }

    /**
     * Invalid: overloaded methods are not in telescoping order.
     * No-arg method should come before one-arg method.
     *
     * @see #setName(String)
     * @see #setName() // violation
     * @see #setName(String, int)
     */
    public void getName(String value) {
    }

    /**
     * Invalid: local member appears after simple class reference.
     *
     * @see OtherClass
     * @see #field // violation
     */
    public void setName() {
    }

    /**
     * Invalid: simple class member appears before simple class reference.
     *
     * @see OtherClass#field
     * @see OtherClass // violation
     */
    public void setName(String value) {
    }

    /**
     * Invalid: simple class method appears before simple class constructor.
     *
     * @see OtherClass#method()
     * @see OtherClass#OtherClass() // violation
     */
    public void setName(String value, int age) {
    }

    /**
     * Invalid: qualified class reference appears before simple class reference.
     *
     * @see java.util.List
     * @see OtherClass // violation
     */
    public void invalidQualifiedBeforeSimpleClass() {
    }

    /**
     * Invalid: package reference appears before qualified class reference.
     *
     * @see java.util
     * @see java.util.List // violation
     */
    public void invalidPackageBeforeQualifiedClass() {
    }

}
```

**Comment by romani** [comment]
in case of uncertainty we should do nothing and let user order as he likes.
We should do violations only when we confident/certain.

**Comment by Carbon14-48** [comment]
@gianmarcoschifone hi i would love to work on this i just have some questions
The issue title says `JavadocSeeOrder`, but the config snippet shows `<module name="`JavadocSeeTagOrder`"/>`. Which should the check be named? 
and should i implment all 7 rules ?  rules 1–5 are unambiguous and safe to land first, while 6–7 are uncertain and causes unambiguity in some edge cases

**Comment by romani** [comment]
JavadocSeeTagOrder is name, title is updated.

6-7 are just rules for overloaded constructors/method.
overloaded method are method that has the same name, should be easy to parse. 
Check should not care is this method or ctor, if identifier is same they should be close to each other and the less parameters is above.

**Comment by romani** [comment]
Fix is merged
